src/runtime_src/core/edge/user/shim.cpp

```diff
@@ -495,13 +503,179 @@ xclLoadXclBin(const xclBin *buffer)
+  }
+
+  return 0;
+}
+
+static int
+copyBufferToFile(std::string file_path, const char* buf, uint64_t size)
+{
+  std::ofstream file(file_path, std::ios::out | std::ios::binary);
+
+  if(!file) {
+    xclLog(XRT_ERROR, "%s: failed to open '%s' for writing xclbin section",__func__, file_path);
+    return -1;
+  }
+  file.write(buf, size);
+  file.close();
+
+  return 0;
+}
+
+static int
+libdfxConfig(std::string& xclbin_dir_path, const axlf *top,
+	     const axlf_section_header *bit_header, const axlf_section_header *overlay_header)
+{
+  // create a temp directory to extract bitstream and dtbo
+  char dir[] = "/tmp/xclbin.XXXXXX";
+  char *tmpdir = mkdtemp(dir);
+  if(tmpdir == nullptr) {
+    xclLog(XRT_ERROR, "%s: failed to create tmp directory for xclbin files extraction",__func__);
+    return -1;
+  }
+
+  xclbin_dir_path = tmpdir;
+  // create a file with BITSTREAM section
+  std::string bit_file_path = xclbin_dir_path + "/xclbin.bit";
+  auto bit_buffer = reinterpret_cast<const char *>(top) + bit_header->m_sectionOffset;
+  if(copyBufferToFile(bit_file_path, bit_buffer, bit_header->m_sectionSize))
+    return -1;
+
+  // create a file with OVERLAY(dtbo) section
+  std::string overlay_file_path = xclbin_dir_path + "/xclbin.dtbo";
+  auto overlay_buffer = reinterpret_cast<const char *>(top) + overlay_header->m_sectionOffset;
+  if(copyBufferToFile(overlay_file_path, overlay_buffer, overlay_header->m_sectionSize))
+    return -1;
+
+  return 0;
+}
+
+// function for cleaning temp files
+static void
+libdfxClean(const std::string& file_path)
+{
+  try {
+    if(boost::filesystem::exists(boost::filesystem::path(file_path)))
+      boost::filesystem::remove_all(boost::filesystem::path(file_path));
+  }
+  catch(std::exception& ex) { /* do nothing */ }
+
+  return;
+}
+
+static int
+libdfxLoadAxlf(std::shared_ptr<xrt_core::device> core_dev, const axlf *top,
+	       const axlf_section_header *overlay_header, int& fd, int flags, std::string& dtbo_path)
+{
+  static const std::string ZOCL_DRM_DEVICE = "/dev/dri/renderD128";
+  static const std::string FPGA_DEVICE = "/dev/fpga0";
+  int ret = 0;
+
+  // check BITSTREAM section
+  const axlf_section_header *bit_header = xclbin::get_axlf_section(top, axlf_section_kind::BITSTREAM);
+  if(!bit_header) {
+    xclLog(XRT_ERROR, "%s: No BITSTREAM section in xclbin", __func__);
+    return -1;
+  }
+
+  //check if xclbin is already loaded
+  try {
+    if(core_dev->get_xclbin_uuid() == xrt::uuid(top->m_header.uuid) && !(flags & DRM_ZOCL_FORCE_PROGRAM)) {
+      xclLog(XRT_INFO, "%s: skipping as xclbin is already loaded", __func__);
+      return 0;
+    }
+  }
+  catch(const std::exception &e) {
+    // can happen when no bitstream is loaded and xclbinid sysfs is not created
+    // do nothing
+  }
+
+  ret = libdfxHelper(core_dev, dtbo_path, fd);
+  if(ret)
+    return ret;
+
+  std::string xclbin_dir_path;
+  ret = libdfxConfig(xclbin_dir_path, top, bit_header, overlay_header);
+  if(ret)
```

Comment:
This is a prime example of why C++ uses exceptions.  For if we were doing so, then we wouldn't need to have this return value checks.